staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -161,29 +162,36 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		defer utilruntime.HandleCrash()
 		<-stopCh
 
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
 		if client != nil {
 			client.Close()
 			clientErr = fmt.Errorf("server is shutting down")
 		}
 	}()
 
+	var requestGroup singleflight.Group
+	const checkETCDKey = "checkETCD"
 	return func() error {
 		// Given that client is closed on shutdown we hold the lock for
 		// the entire period of healthcheck call to ensure that client will
 		// not be closed during healthcheck.
 		// Given that healthchecks has a 2s timeout, worst case of blocking
 		// shutdown for additional 2s seems acceptable.
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
```

Comment:
@wojtek-t @liggitt I think this is a bug, since all healtcheck calls are serialized, this fixes overwhelm etcd, but it should be adding a considerable delay to clients

Reply:
you're saying it was a bug before that would serialize health check calls, and is fixed in this PR, right?

Reply:
yes, every call to the healthcheck holds the lock, thus , all the calls has to be serialized

Reply:
that was true previously, and is no longer true in this PR? just making sure I'm reading this correctly

Reply:
yes, I've added it as a separate commit to differntiate the changes